File: swift/0530. Minimum Absolute Difference in BST.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func getMinimumDifference(_ root: TreeNode?) -> Int {
        var array = [Int]()
        
        helper(root, &array)
        
        return getMinAbsDifference(for: array)
    }
    
    private func helper(_ root: TreeNode?, _ array: inout [Int]) {
        guard let root = root else { return }
        
        array.append(root.val)
        helper(root.left, &array)
        helper(root.right, &array)
    }
                        
    private func getMinAbsDifference(for array: [Int]) -> Int {
        var array = array.sorted()
        var minDifference = Int.max
        
        for i in 1..<array.count {
            minDifference = min(abs(array[i-1] - array[i]), minDifference)
        }
        
        return minDifference
    }
}

File: python/0530. Minimum Absolute Difference in BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
     def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
            def dfsCreateArray(root: Optional[TreeNode], array):
                if root is None: return 
        
                array.append(root.val)
                dfsCreateArray(root.left, array)
                dfsCreateArray(root.right, array)
                
            def getAbsMinDifference(array) -> int:
                array.sort()
                absMin = float('inf')
                for i in range(1, len(array)):
                    difference = abs(array[i-1] - array[i])
                    absMin = min(difference, absMin)
                    
                return absMin
                
            array = []
            dfsCreateArray(root, array)
            return getAbsMinDifference(array)
